File: Greedy/1151. Minimum Swaps to Group All 1's Together/1151. Minimum Swaps to Group All 1's Together.cpp
class Solution {
public:
    int minSwaps(vector<int>& data) {
        int n = data.size();
        vector<int>countzero(n+1, 0);
        int count = 0;
        for (int i = 0; i < n; i++) {
            if (data[i] == 0) count++;
            countzero[i+1] = count;
        }
        int size = n-count;
        //System.out.println(size + " " + count);
        int rets = INT_MAX;
        for (int i = size-1; i < n; i++) {
            rets = min(rets, countzero[i+1] - countzero[i+1-size]);
        }
        return rets;
    }
};

File: TwoPointer/1151. Minimum Swaps to Group All 1's Together/1151. Minimum Swaps to Group All 1's Together.java
class Solution {
    public int minSwaps(int[] data) {
        int n = data.length;
        int count = 0;
        for (int d : data) {
            count += (d == 1) ? 1 : 0;
        }
        if (count < 2) return 0;
        
        int left = 0, right = 0;
        int curcount = 0;
        int rets = Integer.MAX_VALUE;
        for (right = 0; right < n; right++) {
            if (data[right] == 1) curcount++;
            while (left < right && right-left+1 > count) {
                if (data[left] == 1) curcount--;
                left++;
            }
            int len = right-left+1;
            if (count == len) rets = Math.min(rets, len-curcount);
        }
        return rets;
    }
}
